nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DeduplicateRecords.java

```diff
@@ -0,0 +1,690 @@
+            .build();
+
+    static final Relationship REL_NON_DUPLICATE = new Relationship.Builder()
+            .name("non-duplicate")
+            .description("If the record was not found in the cache, it will be routed to this relationship")
+            .build();
+
+    static final Relationship REL_ORIGINAL = new Relationship.Builder()
+            .name("original")
+            .description("The original input FlowFile is sent to this relationship unless there is a fatal error in the processing.")
+            .build();
+
+    static final Relationship REL_FAILURE = new Relationship.Builder()
+            .name("failure")
+            .description("If unable to communicate with the cache, the FlowFile will be penalized and routed to this relationship")
+            .build();
+
+    private List<PropertyDescriptor> descriptors;
+
+    private Set<Relationship> relationships;
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        final List<PropertyDescriptor> descriptors = new ArrayList<>();
+        descriptors.add(DEDUPLICATION_STRATEGY);
+        descriptors.add(DISTRIBUTED_MAP_CACHE);
+        descriptors.add(CACHE_IDENTIFIER);
+        descriptors.add(RECORD_READER);
+        descriptors.add(RECORD_WRITER);
+        descriptors.add(INCLUDE_ZERO_RECORD_FLOWFILES);
+        descriptors.add(RECORD_HASHING_ALGORITHM);
+        descriptors.add(FILTER_TYPE);
+        descriptors.add(FILTER_CAPACITY_HINT);
+        descriptors.add(BLOOM_FILTER_FPP);
+        this.descriptors = Collections.unmodifiableList(descriptors);
+
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_DUPLICATE);
+        relationships.add(REL_NON_DUPLICATE);
+        relationships.add(REL_ORIGINAL);
+        relationships.add(REL_FAILURE);
+        this.relationships = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+        return new PropertyDescriptor.Builder()
+                .name(propertyDescriptorName)
+                .description("Specifies a value to use from the record that matches the RecordPath: '" +
+                        propertyDescriptorName + "' which is used together with other specified " +
+                        "record path values to determine the uniqueness of a record. " +
+                        "Expression Language may reference variables 'field.name', 'field.type', and 'field.value' " +
+                        "to access information about the field and the value of the field being evaluated.")
```

Comment:
Maybe mention that this is dependent on `Single File`, and for the sake of the Usage doc (since it will be moot once a dynamic property is added to the dialog) maybe mention what happens if no dynamic property is provided.

I was also getting some weird behavior with dynamic properties. It looks like if no dynamic properties are provided, all fields are used to determine duplication. If (in my example with 4 records) all records were exactly the same, it sent 1 to non-dupe and 3 to dupe. If none were exactly the same (even if some had individually identical fields like lastName), it sent all to non-dupe. If I set a dynamic property `/lastName` to `Doe` (using your sample data), the `John Smith` record was sent to non-dupe and the others (including the one with lastName = Doe) to dupe. If I awr `/lastName` to `Smith` it did the exact same thing. Only when I set `/lastName` to `${field.value}` did it behave as expected (the first `Smith` and the `Doe` to non-dupe and the second `Smith` to dupe).

Also when I put in a record path for a missing field `/lastname` (lowercased), it behaved as if none had been provided. I wonder if we want to have an error/failure condition there as the user may not be aware that the behavior will not be as expected.

Reply:
> If none were exactly the same (even if some had individually identical fields like lastName), it sent all to non-dupe.

That's expected behavior.

Reply:
You brought up a good point about the EL stuff there. I think that's something I missed when merging @adamfisher's PR with mine. Now that I think about it, I think the better way to do this property stuff would be to do it like this:

<human_readable> => record path

Ex. `firstName => /firstName`

That would make it easier to say "hey, we didn't find firstName" instead of throwing a record path statement (which could be long) into the error logs.

EL should only be supported on the cache identifier. I can't honestly think of a good use case to give users the flexibility to dynamic define property values on this part with EL. Thoughts?

@exceptionfactory what do you think?

Reply:
Supporting EL for the record path property might allow for some use case where a path expression has a dynamic predicate value, but I can also see it being confusing given that it would only work in the context of a single file. Removing the EL support for FlowFile attributes makes sense under the circumstances described.